map/couch/patient.py

```diff
@@ -41,94 +48,170 @@ def dbname_from_username(username):
+        # Add the new couch_db identifier to the patient FHIR
+        self.patient_fhir = update_identifier(
+            self.patient_fhir, self.couch_id())
 
-    If no couch db exists for given patient, a new couch user and matching
-    couch db will be generated, populated with the given patient_fhir.
+        # Add user to couch, which also generates db
+        couch.add_user(name=self.username, password='auto', roles=['patient'])
 
-    Returns potentially modified patient_fhir, if newer version is found
+        # Couch is configured (see ``zz-couch_defaults.ini``) to generate a
+        # per-user database on user creation.  Apparent API race condition
+        # requires the try, try model below.
+        db = None
+        try:
+            current_app.logger.debug("attempt to access new db")
+            db = couch[self.userdbname]
+        except ResourceNotFound:
+            current_app.logger.debug("404 on new user db")
+
+        if not db:
+            try:
+                current_app.logger.debug("attempt to directly create new db")
+                couch.create(self.userdbname)
+                current_app.logger.debug("success")
+            except ServerError as e:
+                assert 'conflict' in str(e)
+            finally:
+                db = couch[self.userdbname]
+
+        # Push identifier upstream
+        HapiRequest.put_resource(self.patient_fhir)
+        # Now persist the given/modified patient document
+        self.sync_document(self.patient_fhir)
+
+    def sync(self):
+        """API to invoke complete sync of patient and related resources"""
+        self.sync_patient()
+        self.sync_related_resources()
+        return self.patient_fhir
+
+    def sync_document(self, document):
+        """sync contents of any given document w/ couch user db
+
+        Document may or may not previously exist in couch.  Guaranteed to
+        exist after call.  If a newer version is found in couch, that will
+        be returned.
+
+        All documents are keyed by ``ResourceType/Id``, such as
+        ``CarePlan/54``
+
+        """
+        key = f"{document['resourceType']}/{document['id']}"
+        db = couch[self.userdbname]
+        if key not in db:
+            db[key] = document
+            return document
+
+        # HAPI maintains last_updated (ISO 8601 format); replace if newer
+        hapi_time = dt_or_none(
+            document.get('meta', {}).get('lastUpdated'))
+        couch_time = dt_or_none(db[key].get('meta', {}).get('lastUpdated'))
+        if (couch_time and not hapi_time) or (couch_time > hapi_time):
+            current_app.logger.debug(
+                f"found newer data in couch for {key}; push to HAPI")
+            document = HapiRequest.put_resource(db[key])
+
+        elif (hapi_time and not couch_time) or (hapi_time > couch_time):
+            current_app.logger.debug(
+                f"found newer data in HAPI for {key}; push to couch")
```

Comment:
Nitpick: [`logger` uses printf-style string formatting builtin](https://reinout.vanrees.org/weblog/2015/06/05/logging-formatting.html)

```suggestion
                "found newer data in HAPI for %s; push to couch", key)
```